File: src/test/java/io/github/rygel/sparkle4j/SignatureVerifierTest.java
package io.github.rygel.sparkle4j;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyPairGenerator;
import java.security.Signature;
import java.util.Arrays;
import java.util.Base64;

@SuppressWarnings("NullAway.Init")
class SignatureVerifierTest {

    @TempDir Path tempDir;

    @Test
    @DisplayName("valid signature returns true")
    void validSignature() throws Exception {
        var keyPair = KeyPairGenerator.getInstance("Ed25519").generateKeyPair();
        var content = "test content".getBytes(StandardCharsets.UTF_8);
        var file = tempDir.resolve("test-file");
        Files.write(file, content);

        var sig = Signature.getInstance("Ed25519");
        sig.initSign(keyPair.getPrivate());
        sig.update(content);
        var signatureBase64 = Base64.getEncoder().encodeToString(sig.sign());

        // Use DER-encoded public key (44 bytes)
        var publicKeyBase64 = Base64.getEncoder().encodeToString(keyPair.getPublic().getEncoded());

        var verifier = new SignatureVerifier(publicKeyBase64);
        assertTrue(verifier.verify(file, signatureBase64));
    }

    @Test
    @DisplayName("tampered file returns false")
    void tamperedFile() throws Exception {
        var keyPair = KeyPairGenerator.getInstance("Ed25519").generateKeyPair();
        var content = "test content".getBytes(StandardCharsets.UTF_8);
        var file = tempDir.resolve("test-file");
        Files.write(file, "tampered content".getBytes(StandardCharsets.UTF_8));

        var sig = Signature.getInstance("Ed25519");
        sig.initSign(keyPair.getPrivate());
        sig.update(content);
        var signatureBase64 = Base64.getEncoder().encodeToString(sig.sign());

        var publicKeyBase64 = Base64.getEncoder().encodeToString(keyPair.getPublic().getEncoded());

        var verifier = new SignatureVerifier(publicKeyBase64);
        assertFalse(verifier.verify(file, signatureBase64));
    }

    @Test
    @DisplayName("invalid base64 key returns false")
    void invalidBase64Key() throws Exception {
        var file = tempDir.resolve("test-file");
        Files.write(file, "content".getBytes(StandardCharsets.UTF_8));

        var verifier = new SignatureVerifier("not-valid-base64!!!");
        assertFalse(verifier.verify(file, Base64.getEncoder().encodeToString(new byte[64])));
    }

    @Test
    @DisplayName("invalid key format returns false")
    void invalidKeyFormat() throws Exception {
        var file = tempDir.resolve("test-file");
        Files.write(file, "content".getBytes(StandardCharsets.UTF_8));

        var verifier = new SignatureVerifier(Base64.getEncoder().encodeToString(new byte[16]));
        assertFalse(verifier.verify(file, Base64.getEncoder().encodeToString(new byte[64])));
    }

    @Test
    @DisplayName("raw 32-byte key format works")
    void raw32ByteKey() throws Exception {
        var keyPair = KeyPairGenerator.getInstance("Ed25519").generateKeyPair();
        var content = "test content".getBytes(StandardCharsets.UTF_8);
        var file = tempDir.resolve("test-file");
        Files.write(file, content);

        var sig = Signature.getInstance("Ed25519");
        sig.initSign(keyPair.getPrivate());
        sig.update(content);
        var signatureBase64 = Base64.getEncoder().encodeToString(sig.sign());

        // Extract raw 32-byte key from DER-encoded public key (last 32 bytes of 44-byte encoding)
        var derEncoded = keyPair.getPublic().getEncoded();
        var rawKey = Arrays.copyOfRange(derEncoded, 12, 44);
        var publicKeyBase64 = Base64.getEncoder().encodeToString(rawKey);

        var verifier = new SignatureVerifier(publicKeyBase64);
        assertTrue(verifier.verify(file, signatureBase64));
    }

    @Test
    @DisplayName("DER-encoded 44-byte key format works")
    void derEncodedKey() throws Exception {
        var keyPair = KeyPairGenerator.getInstance("Ed25519").generateKeyPair();
        var content = "another test".getBytes(StandardCharsets.UTF_8);
        var file = tempDir.resolve("test-file");
        Files.write(file, content);

        var sig = Signature.getInstance("Ed25519");
        sig.initSign(keyPair.getPrivate());
        sig.update(content);
        var signatureBase64 = Base64.getEncoder().encodeToString(sig.sign());

        // Use full 44-byte DER-encoded key
        var publicKeyBase64 = Base64.getEncoder().encodeToString(keyPair.getPublic().getEncoded());

        var verifier = new SignatureVerifier(publicKeyBase64);
        assertTrue(verifier.verify(file, signatureBase64));
    }
}

File: src/test/kotlin/io/github/sparkle4j/SignatureVerifierTest.kt
package io.github.sparkle4j

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Files
import java.nio.file.Path
import java.security.KeyPairGenerator
import java.util.Base64
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class SignatureVerifierTest {

    @TempDir
    lateinit var tempDir: Path

    // Generate a real Ed25519 keypair for tests (Ed25519 needs no explicit initialize)
    private val keyPair = KeyPairGenerator.getInstance("Ed25519").generateKeyPair()

    private val publicKeyBase64: String =
        Base64.getEncoder().encodeToString(keyPair.public.encoded)

    private fun sign(data: ByteArray): String {
        val sig = java.security.Signature.getInstance("Ed25519")
        sig.initSign(keyPair.private)
        sig.update(data)
        return Base64.getEncoder().encodeToString(sig.sign())
    }

    @Test
    fun `verifies valid signature`() {
        val content = "hello sparkle4j".toByteArray()
        val file = tempDir.resolve("update.exe").also { Files.write(it, content) }
        val signature = sign(content)

        val verifier = SignatureVerifier(publicKeyBase64)
        assertTrue(verifier.verify(file, signature))
    }

    @Test
    fun `rejects signature for tampered file`() {
        val original = "original content".toByteArray()
        val signature = sign(original)

        val tampered = "tampered content".toByteArray()
        val file = tempDir.resolve("update.exe").also { Files.write(it, tampered) }

        val verifier = SignatureVerifier(publicKeyBase64)
        assertFalse(verifier.verify(file, signature))
    }

    @Test
    fun `skips verification and returns true when no public key configured`() {
        val file = tempDir.resolve("update.exe").also { Files.write(it, byteArrayOf(1, 2, 3)) }
        val verifier = SignatureVerifier(null)
        // Should warn and return true (verification disabled)
        assertTrue(verifier.verify(file, "any-signature"))
    }

    @Test
    fun `returns false for invalid base64 signature`() {
        val file = tempDir.resolve("update.exe").also { Files.write(it, byteArrayOf(1)) }
        val verifier = SignatureVerifier(publicKeyBase64)
        assertFalse(verifier.verify(file, "not-valid-base64!!!"))
    }

    @Test
    fun `returns false for invalid public key encoding`() {
        val content = "data".toByteArray()
        val file = tempDir.resolve("update.exe").also { Files.write(it, content) }
        val verifier = SignatureVerifier(Base64.getEncoder().encodeToString(byteArrayOf(0, 1, 2)))
        assertFalse(verifier.verify(file, sign(content)))
    }

    @Test
    fun `accepts raw 32-byte Ed25519 public key (Sparkle key format)`() {
        // Extract just the raw 32-byte key from the DER-encoded public key
        val derEncoded = keyPair.public.encoded // 44 bytes (12-byte header + 32-byte key)
        val rawKey32 = derEncoded.takeLast(32).toByteArray()

        val content = "test payload".toByteArray()
        val file = tempDir.resolve("update.exe").also { Files.write(it, content) }
        val signature = sign(content)

        val verifier = SignatureVerifier(Base64.getEncoder().encodeToString(rawKey32))
        assertTrue(verifier.verify(file, signature), "Raw 32-byte key should be accepted")
    }

    @Test
    fun `accepts DER-encoded X509 Ed25519 public key (Java keytool format)`() {
        // Standard Java-format public key (44-byte DER)
        val content = "test payload".toByteArray()
        val file = tempDir.resolve("update.exe").also { Files.write(it, content) }
        val signature = sign(content)

        val verifier = SignatureVerifier(publicKeyBase64) // publicKeyBase64 is DER-encoded (44 bytes)
        assertTrue(verifier.verify(file, signature), "DER X.509 key should be accepted")
    }
}
